src/radical/saga/task.py

```diff
@@ -121,6 +123,34 @@ def __init__ (self, _adaptor, _method_type, _method_context, _ttype) :
             pass
 
 
+    # --------------------------------------------------------------------------
+    #
+    def __eq__ (self, other) :
+
+        # we need to define __eq__ to ensure that tasks and jobs with different
+        # IDs but same dict representation can be added to Python containers
+        # (see Container.add()).
+
+        id_self  = self.get_id()
+        id_other = None
+
+        if other:
+            id_other = other.get_id()
+
+        if id_self and id_other:
+            return id_self == id_other
+        else:
+            return super(Task, self).__eq__(other)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def get_id(self):
+
+        # implement get_id for __eq__
+        return self._uid
+
+
```

Comment:
could we simplify this? ID is required and it is always set
```python
def __eq__(self, other):
    if isinstance(other, Task):
        return self._id == other._id
    else:
        return super(Task, self).__eq__(other)
```

Reply:
```
t = Task()
if (t == None): 
    ...
```
`None` won't have a get_id, and the code would raise on trying to call it.  Or am I missing something?

Reply:
I actually don't see how task-object can be `None`, when it is created it is always the object of `Task` class (`isinstance(t, Task)`). Don't know could it be changed what it returns, but it always will have Task-related methods and attributes
```
>>> import radical.saga as rs
>>> t=rs.Task()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() missing 4 required positional arguments: '_adaptor', '_method_type', '_method_context', and '_ttype'
```

Reply:
Yes, sure, that was supposed to be pseudo code.  The point is, you can compare a task object against other types, and those may not have a `get_id()` method.  Here is a complete example:
```py
#!/usr/bin/env python3

class Foo(object):

    def __init__(self, uid):

        self.uid = uid

    def __eq__(self, other):

        return self.uid == other.uid

class Bar(object):
    pass


f1a = Foo(1)
f1b = Foo(1)
f2a = Foo(2)
b   = Bar()

assert(f1a == f1b)
assert(f1a != f2a)
assert(f1b != f2a)

assert(f1a)
assert(b)

b == f1a  # raises
```

which results in:
```sh
./t.py 
Traceback (most recent call last):
  File "./t.py", line 29, in <module>
    b == f1a
  File "./t.py", line 11, in __eq__
    return self.uid == other.uid
AttributeError: 'Bar' object has no attribute 'uid'
```

Reply:
I agree when it is compared task-object and non-task-object, but our case is different:
```python
# inside Task.__eq__
    if isinstance(other, Task):
        return self.get_id() == other.get_id()
```
(and if `other` is a non-task-object then call an inherited method `super(Task, self).__eq__(other)`, as you proposed)

Reply:
Fair - thanks.